File: san.py
import re
from typing import Tuple, Optional

from models.game import ChessGame
from models.pieces import WHITE, BLACK, ChessPieceType

SAN_REGEX = re.compile(r'^([NBRQK])?([a-h])?([1-8])?[x-]?([a-h][1-8])(=?[NBRQK])?[+#]?$', re.I)

PIECE_MAP = {
    None: ChessPieceType.PAWN,
    "N": ChessPieceType.KNIGHT,
    "B": ChessPieceType.BISHOP,
    "R": ChessPieceType.ROOK,
    "Q": ChessPieceType.QUEEN,
    "K": ChessPieceType.KING,
}

PROMO_MAP = {
    None: None,
    "N": ChessPieceType.KNIGHT,
    "B": ChessPieceType.BISHOP,
    "R": ChessPieceType.ROOK,
    "Q": ChessPieceType.QUEEN,
    "K": ChessPieceType.KING,
}

def parse_san(game: ChessGame, san: str, color: str) -> Tuple[str, str, Optional[ChessPieceType]]:
    """Parse a simple SAN move and return (start, end, promotion)."""
    san = san.replace("0", "O")
    if san.upper() in {"O-O", "O-O+", "O-O#"}:
        start = "e1" if color == WHITE else "e8"
        end = "g1" if color == WHITE else "g8"
        return start, end, None
    if san.upper() in {"O-O-O", "O-O-O+", "O-O-O#"}:
        start = "e1" if color == WHITE else "e8"
        end = "c1" if color == WHITE else "c8"
        return start, end, None
    m = SAN_REGEX.match(san)
    if not m:
        raise ValueError(f"cannot parse SAN: {san}")
    piece_letter, dfile, drank, dest, promo = m.groups()
    piece_type = PIECE_MAP[piece_letter.upper() if piece_letter else None]
    dis_file = dfile
    dis_rank = drank
    promo_piece = PROMO_MAP[promo.strip("=").upper() if promo else None]
    moves = game.board.all_legal_moves(color)
    candidates = []
    for start, ends in moves.items():
        if dest not in ends:
            continue
        piece = game.board[start]
        if piece.type != piece_type:
            continue
        if dis_file and start[0] != dis_file:
            continue
        if dis_rank and start[1] != dis_rank:
            continue
        candidates.append(start)
    if len(candidates) != 1:
        raise ValueError(f"ambiguous SAN: {san}")
    return candidates[0], dest, promo_piece

File: src/san.rs
use crate::{
    game::Game,
    pieces::{Color, PieceType},
};
use regex::Regex;

pub fn parse_san(game: &mut Game, san: &str, color: Color) -> Option<(String, String)> {
    let mut san = san.replace("0", "O");
    san = san.trim_end_matches(|c| c == '+' || c == '#').to_string();
    let upper = san.to_uppercase();
    if upper == "O-O" {
        let start = if color == Color::White { "e1" } else { "e8" };
        let end = if color == Color::White { "g1" } else { "g8" };
        return Some((start.into(), end.into()));
    }
    if upper == "O-O-O" {
        let start = if color == Color::White { "e1" } else { "e8" };
        let end = if color == Color::White { "c1" } else { "c8" };
        return Some((start.into(), end.into()));
    }
    let re = Regex::new(r"^([NBRQK])?([a-h])?([1-8])?[x-]?([a-h][1-8])(=?[NBRQK])?$").ok()?;
    let caps = re.captures(&san)?;
    let mut piece_letter = caps.get(1).map(|m| m.as_str());
    let mut dfile = caps.get(2).map(|m| m.as_str());
    let drank = caps.get(3).map(|m| m.as_str());
    let dest = caps.get(4)?.as_str();
    if let Some(pl) = piece_letter {
        if pl.chars().all(|c| c.is_ascii_lowercase()) && "abcdefgh".contains(pl) && dfile.is_none()
        {
            dfile = Some(pl);
            piece_letter = None;
        }
    }
    let ptype = match piece_letter {
        Some("N") => PieceType::Knight,
        Some("B") => PieceType::Bishop,
        Some("R") => PieceType::Rook,
        Some("Q") => PieceType::Queen,
        Some("K") => PieceType::King,
        None => PieceType::Pawn,
        _ => return None,
    };
    let moves = game.board.all_legal_moves(color);
    let mut candidates = Vec::new();
    for (start, end) in moves {
        if end != dest {
            continue;
        }
        if let Some(piece) = game.board.get(&start) {
            if piece.piece_type != ptype {
                continue;
            }
            if let Some(df) = dfile {
                if &start[0..1] != df {
                    continue;
                }
            }
            if let Some(dr) = drank {
                if &start[1..2] != dr {
                    continue;
                }
            }
            candidates.push(start);
        }
    }
    if candidates.len() == 1 {
        Some((candidates.remove(0), dest.to_string()))
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::game::Game;

    #[test]
    fn castle_kingside() {
        let mut game = Game::new();
        let mv = parse_san(&mut game, "O-O", Color::White).unwrap();
        assert_eq!(mv, ("e1".to_string(), "g1".to_string()));
    }

    #[test]
    fn queen_move() {
        let mut game = Game::new();
        // open path for queen
        game.make_move("d2", "d4");
        game.make_move("d7", "d5");
        let mv = parse_san(&mut game, "Qd3", Color::White).unwrap();
        assert_eq!(mv, ("d1".to_string(), "d3".to_string()));
    }
}
